#158565106 Save new asset category

FelixWambiri opened the pull request:
**What does this PR do?**
This PR makes use of the API endpoint to create new asset categories.

**Description of Task to be completed?**
* Enable users to create new asset categories.
* Notify the user if any of the required fields are not properly filled.

**How should this be manually tested?**
* visit `http://activo-dev.andela.com:3000/settings/categories/new` to create asset categories.

**Any background context you want to provide?**

The API endpoint currently used to create the asset categories is the mock API as the staging API is currently unavailable for use at the moment.

**What are the relevant pivotal tracker stories?**
[#158565106](https://www.pivotaltracker.com/story/show/158565106)

**Screenshots (if appropriate)**
![image](https://user-images.githubusercontent.com/30326444/43008810-a95f6d2e-8c44-11e8-974d-4dcfd35f362f.png)


**Questions:**

xxx

Billmike commented on src/pages/CreateCategoryPage/CreateCategoryPage.test.tsx:
```diff
@@ -8,9 +8,5 @@ import { mount } from 'enzyme';
-
-    expect(mount).toMatchSnapshot();
-  });
+  it('renders a form for creating a new category');
```
No test for this?

Billmike commented on src/pages/CreateCategoryPage/index.tsx:
```diff
@@ -9,82 +12,216 @@ import Attribute from 'components/Attribute';
+   * @param {Event} event
+   * @returns {void}
+   */
+  onSubmit = (event) => {
```
No tests for this?

Billmike commented on src/pages/CreateCategoryPage/index.tsx:
```diff
@@ -9,82 +12,216 @@ import Attribute from 'components/Attribute';
+   * @param {Object} category
+   * @returns {object}
+   */
+  validateData = (data)  => {
```
Did you write a test for this method?

Billmike commented on src/pages/CreateCategoryPage/index.tsx:
```diff
@@ -9,82 +12,216 @@ import Attribute from 'components/Attribute';
+  }
+}
+
+const mapStateToProps = state => ({
```
Write tests for mapStateToProps and mapDispatchToProps for this component

Billmike commented on src/pages/CreateCategoryPage/index.tsx:
```diff
@@ -9,82 +12,216 @@ import Attribute from 'components/Attribute';
+   * @param {Object} error
+   * @returns {boolean} 
+   */ 
+  hasErrors = (errors) => {
```
A lot of uncovered lines on your branch. You need to write tests to handle every logic implemented in your task, and that includes testing your class methods.

missating commented on src/components/Attribute/interfaces.d.ts:
```diff
@@ -2,6 +2,9 @@ export interface AttributeProps {
   enableConfigOptions?: boolean;
   error?: string;
+  name?: string;
+  onChange?: (e: Object) => any;
```
I think using `event` makes this clearer than using `e`

missating commented on src/pages/CreateCategoryPage/index.tsx:
```diff
@@ -9,82 +12,218 @@ import Attribute from 'components/Attribute';
-            error="This field is required."
-          />
+// actions
+import { addNewAssetCategory } from '../../store/modules/categories/index';
```
We already setup alias in our `webpack config` to resolve files so you don't have to do `'../../store/modules/categories/index'` when importing.

Do this instead `'modules/categories'`

missating commented on src/pages/CreateCategoryPage/index.tsx:
```diff
@@ -9,82 +12,218 @@ import Attribute from 'components/Attribute';
-            placeholder="Asset tag"
-          />
+// interfaces
+import { Errors, NewCategoryProps } from '../../store/modules/categories/interfaces';
```
Change this too.

missating commented on src/store/modules/categories/index.ts:
```diff
@@ -1,8 +1,51 @@
+
+// helper functions
+import { authService } from '../../../utils/auth';
 
```
The alias thing applies to both line 26 and 29

missating commented on src/store/modules/categories/index.ts:
```diff
@@ -1,8 +1,51 @@
+    'x-mock-response-code': 201,
+    ContentType: 'application/json',
+  },
+});
```
Extract this `axios` instance into a helper function.

The baseURL should be in the env file too

missating commented on src/store/modules/categories/index.ts:
```diff
@@ -21,19 +61,64 @@ export const getAssetCategoriesFailure = (): CategoriesActionFailure => {
+  .then((response) => {
+    dispatch(addAssetCategorySuccess(response.data));
+    dispatch(displayToastMessage('Asset Category successfully created', 'success'));
+    history.push('/dashboard');
```
Yeah. I agree with you. @Billmike

kevinejiro commented:
The pivotal tracker ID in your PR should be a link to the actual story on the PT board.

missating commented:
@FelixWambiri Why do you have 55 files changed? I'm very much surprised. :tired_face:

missating commented on config/webpack.dev.js:
```diff
@@ -4,7 +4,7 @@ const path = require('path');
   hotModuleReplacementPlugin
-} = require('./webpack.plugins');
+} = require('./webpack.plugins.js');
 
```
I don't think there is any need for this change, in our `webpack config` we are already resolving file extensions including the `.js`, you don't need to add it when importing again. 

Please revert this change.

missating commented on nightwatchGlobalSetup.js:
```diff
@@ -1,6 +1,6 @@
 
-import app from './app';
+import app from 'app';
 
```
Where exactly would `app` be imported from? 

Please revert this too.

missating commented on server.js:
```diff
@@ -1,3 +1,3 @@
-const app = require('./app')
+const app = require('app')
 
```
The `app` comment applies here too

missating commented on src/components/AddAssetForm/index.tsx:
```diff
@@ -11,7 +11,7 @@ import RadioButtons from 'components/RadioButtons';
 // interfaces
-import { AddAssetFormProps } from './interfaces';
+import { AddAssetFormProps } from 'components/AddAssetForm/interfaces';
 
```
This `interface` file belongs in the same folder as the `AddAssetForm` component. There absolutely no need for you to use `components/AddAssetForm/interfaces` since its the same directory. 

Please revert here too and anywhere else this may apply.

missating commented on src/components/AddAssetPane/AddAssetPane.test.tsx:
```diff
@@ -6,7 +6,7 @@ import { mount, shallow } from 'enzyme';
 // components
-import AddAssetPane from './';
+import AddAssetPane from 'components/AddAssetPane';
 
```
The former is better than your change.

This applies to other test files too

FelixWambiri commented:
@missating I changed the way I had done importation of files in one component and this automatically updated in other components but I have reverted the changes in the other files.
